compositor_render/src/transformations/shader/validation.rs

```diff
@@ -0,0 +1,418 @@
+        | naga::TypeInner::Matrix { .. }
+        | naga::TypeInner::Atomic { .. }
+        | naga::TypeInner::Image { .. }
+        | naga::TypeInner::Sampler { .. }
+        | naga::TypeInner::AccelerationStructure
+        | naga::TypeInner::RayQuery
+        | naga::TypeInner::ValuePointer { .. } => {
+            if ti1 != ti2 {
+                return Err(TypeEquivalenceError::TypeStructureMismatch(
+                    type1.inner.clone(),
+                    type2.inner.clone(),
+                ));
+            }
+        }
+
+        naga::TypeInner::Array {
+            base: base1,
+            size: size1,
+            stride: stride1,
+        } => {
+            if let naga::TypeInner::Array {
+                base: base2,
+                size: size2,
+                stride: stride2,
+            } = ti2
+            {
+                if stride1 != stride2 {
+                    return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+                }
+
+                array_size_equivalent(size1, mod1, size2, mod2)?;
+                return type_equivalent(base1, mod1, base2, mod2);
+            }
+            return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+        }
+
+        naga::TypeInner::BindingArray {
+            base: base1,
+            size: size1,
+        } => {
+            if let naga::TypeInner::BindingArray {
+                base: base2,
+                size: size2,
+            } = ti2
+            {
+                array_size_equivalent(size1, mod1, size2, mod2)?;
+                return type_equivalent(base1, mod1, base2, mod2);
+            }
+            return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+        }
+
+        naga::TypeInner::Struct {
+            members: ref members1,
+            span: span1,
+        } => {
+            if let naga::TypeInner::Struct {
+                members: ref members2,
+                span: span2,
+            } = ti2
+            {
+                if span1 != span2 || members1.len() != members2.len() {
+                    return Err(TypeEquivalenceError::TypeStructureMismatch(
+                        ti1.clone(),
+                        ti2.clone(),
+                    ));
+                }
+
+                for (m1, m2) in members1.iter().zip(members2.iter()) {
+                    if m1.binding != m2.binding || m1.name != m2.name || m1.offset != m2.offset {
+                        return Err(TypeEquivalenceError::TypeStructureMismatch(ti1, ti2));
+                    }
+
+                    type_equivalent(m1.ty, mod1, m2.ty, mod2)?;
+                }
+            } else {
+                return Err(TypeEquivalenceError::TypeStructureMismatch(
+                    ti1.clone(),
+                    ti2.clone(),
+                ));
+            }
+        }
+
+        naga::TypeInner::Pointer { .. } => {
+            panic!("this should never happen bc of canonicalization")
```

Comment:
```suggestion
            unreachable!("this should never happen bc of canonicalization")
```

Reply:
I'm afraid to do that, bc executing the unreachable! macro is UB and I'm not 100% certain canonicalization makes this impossible